zebra-chain/src/serialization/compact_size.rs

```diff
@@ -0,0 +1,350 @@
+//! Types and serialization for the Bitcoin `CompactSize` format.
+//!
+//! Zebra decodes `CompactSize` fields into two different types,
+//! depending on the consensus rules that apply to that type:
+//! - [`CompactSizeMessage`] for sizes that must be less than the network message limit, and
+//! - [`CompactSize64`] for flags, arbitrary counts, and sizes that span multiple blocks.
+
+use std::convert::{TryFrom, TryInto};
+
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+
+use crate::serialization::{
+    SerializationError, ZcashDeserialize, ZcashDeserializeInto, ZcashSerialize,
+    MAX_PROTOCOL_MESSAGE_LEN,
+};
+
+/// A compactSize-encoded field that is limited to [`MAX_PROTOCOL_MESSAGE_LEN`].
+/// Used for sizes or counts of objects that are sent in network messages.
+///
+/// # Security
+///
+/// Deserialized sizes must be validated before being used to allocate memory.
+///
+/// Preallocating vectors using untrusted `CompactSize`s allows memory
+/// denial of service attacks. Valid sizes must be less than
+/// `MAX_PROTOCOL_MESSAGE_LEN / min_serialized_item_bytes` (or a lower limit
+/// specified by the Zcash consensus rules or Bitcoin network protocol).
+///
+/// As a defence-in-depth for memory preallocation attacks,
+/// Zebra rejects sizes greater than the protocol message length limit.
+/// (These sizes should be impossible, because each array items takes at
+/// least one byte.)
+///
+/// # Serialization Examples
+///
+/// ```
+/// use zebra_chain::serialization::{CompactSizeMessage, ZcashSerialize, MAX_PROTOCOL_MESSAGE_LEN};
+/// use std::convert::TryFrom;
+///
+/// let size = CompactSizeMessage::try_from(0x12).unwrap();
+/// let buf = size.zcash_serialize_to_vec().unwrap();
+/// assert_eq!(buf, b"\x12");
+///
+/// let size = CompactSizeMessage::try_from(0xfd).unwrap();
+/// let buf = size.zcash_serialize_to_vec().unwrap();
+/// assert_eq!(buf, b"\xfd\xfd\x00");
+///
+/// let size = CompactSizeMessage::try_from(0xaafd).unwrap();
+/// let buf = size.zcash_serialize_to_vec().unwrap();
+/// assert_eq!(buf, b"\xfd\xfd\xaa");
+///
+/// let max_size = u64::try_from(MAX_PROTOCOL_MESSAGE_LEN).unwrap();
+/// let size = CompactSizeMessage::try_from(max_size).unwrap();
+/// let buf = size.zcash_serialize_to_vec().unwrap();
+/// assert_eq!(buf, b"\xfe\x00\x00\x20\x00");
+/// ```
+///
+/// [`CompactSizeMessage`]s greater than the maximum network message length
+/// can not be constructed:
+/// ```
+/// # use zebra_chain::serialization::{CompactSizeMessage, MAX_PROTOCOL_MESSAGE_LEN};
+/// # use std::convert::TryFrom;
+/// # let max_size = u64::try_from(MAX_PROTOCOL_MESSAGE_LEN).unwrap();
+/// let err = CompactSizeMessage::try_from(max_size + 1).unwrap_err();
+///
+/// let err = CompactSizeMessage::try_from(0xbbaafd).unwrap_err();
+///
+/// let err = CompactSizeMessage::try_from(0x22ccbbaafd).unwrap_err();
```

Comment:
Nit/Suggestion: Maybe use `assert!` here?

```suggestion
/// assert!(CompactSizeMessage::try_from(max_size + 1).is_err());
///
/// assert!(CompactSizeMessage::try_from(0xbbaafd).is_err());
///
/// assert!(CompactSizeMessage::try_from(0x22ccbbaafd).is_err());
```